Fix regaining control after a remote cursor movement

logiclrd opened the pull request:
I have encountered a bug related to remote cursor movement. When a remote cursor movement is detected, the `cursorModel` is updated so that `gotMouseControl` is `false`. The `_checkPeerControlProtected` method in input_model.dart includes logic to allow you to regain control if the cursor moves more than a certain distance. However, this logic doesn't work as expected because after each check, it latches the new mouse position. As a result, the mouse movement isn't cumulative. The cursor must move `kMouseControlDistance` pixels _in a single event_ in order to regain control. It turns out, at least on my Windows 10 machine, that it takes considerable effort to move the mouse fast enough to trigger this. So what ends up happening is that input is just locked until the cursor leaves the RustDesk remote view window. Then, later, when the mouse cursor re-enters the view, it's usually more than `kMouseControlDistance` pixels away from where it left the window so the `selfGetControl` logic finally triggers.

This PR fixes the bug by removing the line that latches each new mouse event location every time the `selfGetControl` test fails. As a result, now once the mouse has moved a _cumulative_ `kMouseControlDistance` pixels, control is regained.

In addition, the `selfGetControl` calculation uses an imprecise model that checks the distance moved on the X and Y axes separately. As such, it is possible to move the mouse up to 1.414 (sqrt(2)) times `kMouseControlDistance` before `selfGetControl` activates, if the movement is diagonal. This PR updates the calculation to compute the actual distance travelled, so that it triggers after `kMouseControlDistance` in any direction, not just orthogonal directions. (The comparison is done between the square of the distance and the square of the threshold, to avoid performing a square root computation. On any computer less than 30 years old I'm fairly certain this optimization is beyond unnecessary, but it's the principle of the matter. 😛 In another PR, Copilot actually told me to use this exact optimization in another context earlier today, for what it's worth.)

logiclrd committed:
Removed line that erroneously latches the mouse position on failed self get control checks.

logiclrd committed:
Updated the math used to compute selfGetControl to compute the actual distance rather than checking each axis separately.

fufesou commented on flutter/lib/models/input_model.dart:
```diff
@@ -1213,13 +1213,18 @@ class InputModel {
       if (selfGetControl) {
         cursorModel.gotMouseControl = true;
       } else {
-        lastMousePos = ui.Offset(x, y);
```
This is intented.


>  local control is only regained if `kMouseControlDistance` is exceeded in a single mouse event.

As you guessed,
the user on the controlling side has to move the mouse a significant amount (in a short time) to take control.

logiclrd committed:
Added mouse_speed_analyzer.dart.
Updated input_model.dart to import mouse_speed_analyzer.dart and to incorporate an instance into InputModel.
Updated the _checkPeerControlProtected method of InputModel to use the MouseSpeedAnalyzer to track mousement and determine if/when it exceeds the threshold to regain control.

rustdesk commented:
Sorry, I still prefer old regain.

logiclrd commented:
Huh, okay. The current implementation literally doesn't work for me, and it behaves wildly differently depending on the system it's running on (type of mouse). Oh well :-(
